rclpy/rclpy/impl/logger_rcutils.py

```diff
@@ -0,0 +1,275 @@
+    determining the log condition and add them to the context.
+    """
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        # Store all parameters in the context so we can check that users never try to change them.
+        for param in cls.params:
+            context[param] = kwargs.get(param, cls.params[param])
+            if context[param] is None:
+                raise RuntimeError(
+                    'Required parameter "{0}" was not specified for logging filter "{1}"'
+                    .format(param, cls.__name__))
+
+    """
+    Decide if it's appropriate to log given a context, and update the context accordingly.
+    """
+    @staticmethod
+    def log_condition(context):
+        return True
+
+
+class Once(LoggingFilter):
+    """Ignore all log calls except the first one."""
+
+    params = {
+        'once': None,
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
+        context['has_been_logged_once'] = False
+
+    @staticmethod
+    def log_condition(context):
+        logging_condition = False
+        if not context['has_been_logged_once']:
+            logging_condition = True
+            context['has_been_logged_once'] = True
+        return logging_condition
+
+
+class Throttle(LoggingFilter):
+    """Ignore log calls if the last call is not longer ago than the specified duration."""
+
+    params = {
+        'throttle_duration_sec': None,
+        'throttle_time_source_type': 'RCUTILS_STEADY_TIME',
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
+        context['throttle_last_logged'] = 0
+
+    @staticmethod
+    def log_condition(context):
+        logging_condition = True
+        # TODO(dhood): use rcutils time and the time source type
+        now = time.time()
+        next_log_time = context['throttle_last_logged'] + context['throttle_duration_sec']
+        logging_condition = now >= next_log_time
+        if logging_condition:
+            context['throttle_last_logged'] = now
+        return logging_condition
+
+
+class SkipFirst(LoggingFilter):
+    """Ignore the first log call but process all subsequent calls."""
+
+    params = {
+        'skip_first': None,
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
+        context['first_has_been_skipped'] = False
+
+    @staticmethod
+    def log_condition(context):
+        logging_condition = True
+        if not context['first_has_been_skipped']:
+            logging_condition = False
+            context['first_has_been_skipped'] = True
+        return logging_condition
+
+
+# The ordering of this dictionary matches the order in which filters will be processed.
+# Note(dhood): ordering in the constructor is only maintained as of Python 3.6,
```

Comment:
You could initialize the ordered dict in the constructor in older Python versions, e.g. `OrderedDict([('foo', 1), ('bar', 2), ('baz', 3)])`. The initialization in the following lines is nicer to read though. Therefore I would just remove this comment.

Reply:
ah, it was just an issue with how I was using it (I tried to initialised OrderedDict passing a dict, not a list of tuples). I removed the comment in acf11fcf5c98022b9dd0a4bd2703af7ae2fb6183